### ISSUE TITLE
Getting "Database is locked" ONLY in DP_xyz method when using csla&sqlite&entityframework code first?

### ISSUE THREAD

**Issue by StephenChou1017** [MarimerLLC/cslaforum#271]
Hi ：
      Recently I‘m working on a client-side app which use sqlite as the database. I use the csla 4.5.10 version and sqlite entity framework to implement the business logic layer and the data access layer. This works fine in query scenario, but I got the  "Database is locked" errors in the write data scenario.
      This is the unit test code:
     ` [TestMethod]
        public void TestMethod3()
        {
            try
            {
                var newStudent = Student.NewStudent();
                newStudent.FirstName = "Kobe";
                newStudent.LastName = "Bryant";
                newStudent.Id = 11;

                **//Invoke using reflection will work**
                //var mi = newStudent.GetType().GetMethod("AddOne");
                //mi.Invoke(newStudent, null);

                **//Using sqlite entity framework UoW pattern directly here  will work too.**
                //using (var unitOfWork = new DAL.DalFactory().GetUnitOfWork())
                //{
                //    DAL.Student item = new SqliteDemo.Data.Student() { Id = 1, FirstName = "AAA", LastName = "BBB" };
                //    unitOfWork.Set<DAL.Student>().Add(item);
                //    unitOfWork.SaveChanges();
                //}

                **//This will NOT work, which means using sqlite entity framework UoW pattern in DataPortal_Insert will NOT work. Why and how to fix it? ANY REPLY WILL BE APPRECIATED**
                newStudent.Save();

                //var studentInfoList = StudentInfoList.GetAll();
                //Assert.IsNotNull(studentInfoList);
                //Assert.AreEqual<int>(studentInfoList.Count, 11);                
            }
            catch (DataPortalException ex)
            {
                Console.WriteLine(ex.GetBaseException().Message);
            }
`
    Student.AddOne method code in BusinessBase<Student>：
    `public void AddOne()
        {
            using (var unitOfWork = new DAL.DalFactory().GetUnitOfWork())
            {
                DAL.Student item = GetMapDto();
                unitOfWork.Set<DAL.Student>().Add(item);
                unitOfWork.SaveChanges();
            }
        }
`
    Student.DataPortal_Insert method code in BusinessBase<Student>:
    `[Transactional(TransactionalTypes.TransactionScope)]
        protected override void DataPortal_Insert()
        {
            bool cancel = false;
            OnInserting(ref cancel);
            if (cancel) return;

            **//Using sqlite ado.net is OK!!**
            //using (SQLiteConnection c = new SQLiteConnection(@"Data Source=D:\SqliteDB\Coursera.sqlite"))
            //{
            //    c.Open();
            //    using (SQLiteCommand cmd = new SQLiteCommand("INSERT INTO Students (Id, FirstName, LastName, CourseId) VALUES   (11, 'Wendy', 'Wendy', NULL)", c))
            //    {
            //        cmd.ExecuteNonQuery();
            //    }
            //}

            **//Using Sqlite Entity framework is NOT OK.** 
            //SqliteDemoModel context = new SqliteDemoModel();
            //context.Students.Add(GetMapDto());                //A "Database is locked" exception is thrown here, why and how to fix it????
            //context.SaveChanges();

            **//Using Sqlite Entity framework UoW pattern is NOT OK yet.  A "Database is locked" exception is thrown , why and how to fix it????**
            // Get Student data for a single criteria
            //using (var unitOfWork = new DAL.DalFactory().GetUnitOfWork())
            //{
            //    DAL.Student item = GetMapDto();
            //    unitOfWork.Set<DAL.Student>().Add(item);
            //    unitOfWork.SaveChanges();
            //    LoadProperty(IdProperty, item.Id);
            //    OnInserted();
            //}
        }
`
Was I wrong in anywhere ? Any reply will be appreciated! ! You can get the source code from [here](https://github.com/StephenChou1017/CslaSqlite)

**Comment by jonnybee** [comment]
Some basic issues with your code. 

SystemData.Sqlite does not support migrations. So you should either add another NuGet package or disable migrations. Here ex of how to disable: 

    protected override void OnModelCreating(DbModelBuilder modelBuilder)
    {
        Database.SetInitializer<SqliteDemoModel>(null);
    }

You should laso make sure to always have using on the EF Context (or use Csla.EF6 data helper to contain the context for the duration of your call). This may be required so that the context will close connections.

    using (var context = new SqliteDemoModel())
    {
        context.Students.Add(GetMapDto());
        context.SaveChanges();
    }

And finally - I encourage you to add this line to see the actual traffic going to the database:

    public SqliteDemoModel()
        : base("name=SqliteDemoModel")
    {
        Database.Log = (m) => Debug.Print(m);
    }

Sample runs fine with these modifications (and the SQLite.CodeFirst NuGet package to create database).

**Comment by StephenChou1017** [comment]
@jonnybee Thank you for your reply. It runs fine after I add this code:
`Database.SetInitializer<SqliteDemoModel>(null);`

I don't intend to use distributed transactions in sqlite. I want to save two (or more) root class in transactions by using the code: 
`[Transactional(TransactionalTypes.TransactionScope)]`
Then in a Command class I could do like this:
`[Transactional(TransactionalTypes.TransactionScope)]
        protected override void DataPortal_Execute()
        {
            using (var unitOfWork = new DAL.DalFactory().GetUnitOfWork())
            {
                P1.Save(this.ForceUpdate);      //P1 is a BusinessBase object
                P2.Save(this.ForceUpdate);      //P2 is another BusinessBase object, P1 P2 should save both or neither
            }
        }`
In SqlServer, the Csla.Data.DbContextManager<C>(Used in the DalFactory class) will manage and reuse the same connection automatically in transactions, so I thought it will do the same thing in sqlite, but I failded. How can I achieve this in sqlite?  Would you please help to reply once more?

About the database creation and  migrations, originally I intend to adopt this solutions: [http://hintdesk.com/sqlite-with-entity-framework-code-first-and-migration/](http://hintdesk.com/sqlite-with-entity-framework-code-first-and-migration/), now I'll try the Sqlite.CodeFirst NuGet package. 

In short, your answer is more than I expected. Thank you :)

**Comment by jonnybee** [comment]
You should use the DbContextManager to define the lifetime scope in the "rooot" DataPortal_XYZ like this:

     [Transactional(TransactionalTypes.TransactionScope)]
    protected override void DataPortal_Insert()
    {
        bool cancel = false;
        OnInserting(ref cancel);
        if (cancel) return;

        using (var cm= DbContextManager<SqliteDemoModel>.GetManager())
        {
            cm.DbContext.Students.Add(GetMapDto());
 
           // may call othe child dataportals or make other calls 
           // within this lifetimescope of the DbContext

            cm.DbContext.SaveChanges();
        }
    } 

and in the Child objects you would typically have: 

        protected override void Child_Insert()
        {
            using (var cm= DbContextManager<SqliteDemoModel>.GetManager())
            {
                // no need to pass a reference to the DbContext or DbContextManager 
                // The same object is returned within the lifetime scope
                // DbContextManager uses reference counting and will only 
                // call Dispose when referencecount is 1.

                // Add or update child objects here 
            }
        }

**Comment by StephenChou1017** [comment]
@jonnybee Thanks for your reply again.

Based on your reply, I update my code and it works fine in both scenarios: save a root object, or save two (or more) root objects in transactions.  The code is like this:

In the root DataPortal_XYZ:
`[Transactional(TransactionalTypes.TransactionScope)]
        protected override void DataPortal_Insert()
        {
            using (var cm = DbContextManager<SqliteDemoModel>.GetManager())
            {
                DAL.Student item = GetMapDto();
                cm.DbContext.Students.Add(item);

                //The SaveChanges will not run when I save more than one root class in transactions
                if(cm.RefCount == 1)
                    cm.DbContext.SaveChanges();

                LoadProperty(IdProperty, item.Id);
                OnInserted();
            }
        }`

In the Command DataPortal_Execute:
`[Transactional(TransactionalTypes.TransactionScope)]
        protected override void DataPortal_Execute()
        {
            using (var cm = DbContextManager<SqliteDemoModel>.GetManager())
            {
                //P1 P2 are two root objects, they should save both or neither
                P1.Save(this.ForceUpdate);
                P2.Save(this.ForceUpdate);
                
                //Call the SaveChanges once at last.
                cm.DbContext.SaveChanges();
            }
        }`
